tests/components/zha/test_discover.py

```diff
@@ -30,45 +66,298 @@
+    multistate_ch = _ch_mock(zigpy.zcl.clusters.general.MultistateInput)
+
+    class QuirkedIAS(zigpy.quirks.CustomCluster, zigpy.zcl.clusters.security.IasZone):
+        pass
+
+    ias_ch = _ch_mock(QuirkedIAS)
+
+    class _Analog(zigpy.quirks.CustomCluster, zigpy.zcl.clusters.general.AnalogInput):
+        pass
+
+    analog_ch = _ch_mock(_Analog)
+
+    ch_pool = mock.MagicMock(spec_set=zha_channels.ChannelPool)
+    ch_pool.unclaimed_channels.return_value = [
+        door_ch,
+        cover_ch,
+        multistate_ch,
+        ias_ch,
+        analog_ch,
+    ]
+
+    disc.ProbeEndpoint().discover_by_cluster_id(ch_pool)
+    assert probe_mock.call_count == len(ch_pool.unclaimed_channels())
+    probes = (
+        (zha_const.LOCK, door_ch),
+        (zha_const.COVER, cover_ch),
+        (zha_const.SENSOR, multistate_ch),
+        (zha_const.BINARY_SENSOR, ias_ch),
+        (zha_const.SENSOR, analog_ch),
+    )
+    for call, details in zip(probe_mock.call_args_list, probes):
+        component, ch = details
+        assert call[0][0] == component
+        assert call[0][1] == ch
+
+
+def test_single_input_cluster_device_class():
+    """Test SINGLE_INPUT_CLUSTER_DEVICE_CLASS matching by cluster id or class."""
+    _test_single_input_cluster_device_class()
+
+
+def test_single_input_cluster_device_class_by_cluster_class():
+    """Test SINGLE_INPUT_CLUSTER_DEVICE_CLASS matching by cluster id or class."""
+    mock_reg = {
+        zigpy.zcl.clusters.closures.DoorLock.cluster_id: zha_const.LOCK,
+        zigpy.zcl.clusters.closures.WindowCovering.cluster_id: zha_const.COVER,
+        zigpy.zcl.clusters.general.AnalogInput: zha_const.SENSOR,
+        zigpy.zcl.clusters.general.MultistateInput: zha_const.SENSOR,
+        zigpy.zcl.clusters.security.IasZone: zha_const.BINARY_SENSOR,
+    }
+
+    with mock.patch.dict(
+        zha_regs.SINGLE_INPUT_CLUSTER_DEVICE_CLASS, mock_reg, clear=True
+    ):
+        _test_single_input_cluster_device_class()
+
+
+@pytest.mark.parametrize(
+    "override, entity_id",
+    [
+        (None, "light.manufacturer_model_77665544_level_light_color_on_off"),
+        ("switch", "switch.manufacturer_model_77665544_on_off"),
+    ],
+)
+async def test_device_override(hass, zigpy_device_mock, setup_zha, override, entity_id):
+    """Test device discovery override."""
+
+    zigpy_device = zigpy_device_mock(
+        {
+            1: {
+                "device_type": 258,
+                "endpoint_id": 1,
+                "in_clusters": [0, 3, 4, 5, 6, 8, 768, 2821, 64513],
+                "out_clusters": [25],
+                "profile_id": 260,
+            }
+        },
+        "00:11:22:33:44:55:66:77",
+        "manufacturer",
+        "model",
+    )
+
+    if override is not None:
+        override = {"device_config": {"00:11:22:33:44:55:66:77-1": {"type": override}}}
+
+    await setup_zha(override)
+    assert hass.states.get(entity_id) is None
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5051600)